NotificationSystem.Tests/EmailSenderTests.cs

```diff
@@ -0,0 +1,232 @@
+            Mock<ISendGridClient> sendgridMock = new Mock<ISendGridClient>();
+            sendgridMock.Setup(x => x.SendEmailAsync(It.Is<SendGridMessage>(x => x.From == emailFrom &&
+                x.Personalizations.First().Tos.First() == emailTo),
+                It.IsAny<CancellationToken>())).Verifiable();
+
+            EmailSender emailSender = new EmailSender(sendgridMock.Object, emailFromString);
+
+            // Act + Assert
+            Assert.Throws<NullReferenceException>(() => emailSender.ExecuteTask());
+        }
+
+        [Test]
+        public void ValidatePayload_PayloadNotParsed_Exception()
+        {
+            // Arrange
+            string emailFromString = "test@gmail.com";
+            EmailSender emailSender = new EmailSender(null, emailFromString);
+
+            // Act + Assert
+            Assert.Throws<NullReferenceException>(() => emailSender.ValidatePayload());
+        }
+
+
+        [Test]
+        public void ValidatePayload_NonEmptyValues_True()
+        {
+            // Arrange
+            EmailNotification notification = new EmailNotification("test@example.com", "plain text content");
+            string payload = Newtonsoft.Json.JsonConvert.SerializeObject(notification);
+            string emailFromString = "test@gmail.com";
+            EmailSender emailSender = new EmailSender(null, emailFromString);
+
+            // Act
+            emailSender.ParsePayload(payload);
+            bool result = emailSender.ValidatePayload();
+
+            // Assert
+            Assert.AreEqual(true, result);
+        }
+
+        [Test]
+        public void ValidatePayload_EmptyRecipientEmail_False()
+        {
+            // Arrange
+            EmailNotification notification = new EmailNotification("", "plain text content");
+            string payload = Newtonsoft.Json.JsonConvert.SerializeObject(notification);
+            string emailFromString = "test@gmail.com";
+            EmailSender emailSender = new EmailSender(null, emailFromString);
+
+            // Act
+            emailSender.ParsePayload(payload);
+            bool result = emailSender.ValidatePayload();
+
+            // Assert
+            Assert.AreEqual(false, result);
+
+        }
+
+        [Test]
+        public void ValidatePayload_EmptyTextContent_False()
+        {
+            // Arrange
+            EmailNotification notification = new EmailNotification("test@gmail.com", "");
+            string payload = Newtonsoft.Json.JsonConvert.SerializeObject(notification);
+            string emailFromString = "test@gmail.com";
+            EmailSender emailSender = new EmailSender(null, emailFromString);
+
+            // Act
+            emailSender.ParsePayload(payload);
+            bool result = emailSender.ValidatePayload();
+
+            // Assert
+            Assert.AreEqual(false, result);
+        }
+
+        [Test]
+        public void EmailSending_WithValidApiKey_ValidEmail()
```

Comment:
Require sendgrid API key, may be unclear. I think it would be useful to add a comment here.